### PR TITLE
Add concurrency to the GitHub workflows to auto-cancel older runs of actions in PRs.

### PR DESCRIPTION
**Pull request by langleyd**
Resolves https://github.com/vector-im/element-android/issues/5327

### CODE REVIEW

File: .github/workflows/quality.yml [hunk 1]
```diff
@@ -18,6 +18,10 @@ jobs:
   ktlint:
     name: Kotlin Linter
     runs-on: ubuntu-latest
+    # Allow all jobs on main and develop. Just one per PR.
+    concurrency:
+      group: ${{ github.ref == 'refs/heads/main' && format('klint-main-{0}', github.sha) || github.ref == 'refs/heads/develop' && format('klint-develop-{0}', github.sha) || format('klint-{0}', github.ref) }}
```
Comment (michaelkaye) [inline .github/workflows/quality.yml:23]:
```suggestion
      group: ${{ github.ref == 'refs/heads/main' && format('ktlint-main-{0}', github.sha) || github.ref == 'refs/heads/develop' && format('ktlint-develop-{0}', github.sha) || format('ktlint-{0}', github.ref) }}
```
s/klint/ktlint/ to match job name

File: .github/workflows/quality.yml [hunk 2]
```diff
@@ -116,6 +124,10 @@ jobs:
       fail-fast: false
       matrix:
         target: [ Gplay, Fdroid ]
+    # Allow all jobs on develop. Just one per PR.
+    concurrency:
+      group: ${{ github.ref == 'refs/heads/develop' && format('apk-lint-develop-{0}', github.sha) || format('apk-lint-{0}-{1}', matrix.target, github.ref) }}
```
Comment (michaelkaye) [inline .github/workflows/quality.yml:129]:
```suggestion
      group: ${{ github.ref == 'refs/heads/develop' && format('apk-lint-develop-{0}-{1}', matrix.target, github.sha) || format('apk-lint-{0}-{1}', matrix.target, github.ref) }}
```

We want to concurrently run Gplay and Fdroid in develop - they'll both have the same commit hash.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -15,6 +15,10 @@ jobs:
   unit-tests:
     name: Run Unit Tests
     runs-on: ubuntu-latest
+    # Allow all jobs on main and develop. Just one per PR.
+    concurrency:
+      group: ${{ github.ref == 'refs/heads/main' && format('unit-test-main-{0}', github.sha) || github.ref == 'refs/heads/develop' && format('unit-test-develop-{0}', github.sha) || format('unit-test-{0}', github.ref) }}
```
Comment (michaelkaye) [inline .github/workflows/tests.yml:20]:
```suggestion
      group: ${{ github.ref == 'refs/heads/main' && format('unit-tests-main-{0}', github.sha) || github.ref == 'refs/heads/develop' && format('unit-tests-develop-{0}', github.sha) || format('unit-tests-{0}', github.ref) }}
```

s/unit-test/unit-tests/ to match job name.

**Review by michaelkaye — CHANGES_REQUESTED** [review]
Intent of PR is great, looks good.

A few nits around naming (to keep consistency), but I think we have a correctness issue with the missing `matrix.target`

Also wondering if we should add comments for the other workflows as well (Even if it just says "sanity_test.yml: run every time") so we are reminded to consider it in future.

File: .github/workflows/quality.yml (hunk 2 above)
Comment (langleyd) [reply .github/workflows/quality.yml:129]: Ah you're dead right, good catch.

**Comment by langleyd** [comment]
Nice suggestions @michaelkaye , thanks for the review!

**Review by michaelkaye — APPROVED** [review]
This looks good to me.

**Comment by ouchadam** [comment]
out of interest, what's the reasoning for needing more than the head ref? `group: ${{ github.head_ref }}`

**Comment by michaelkaye** [comment]
> out of interest, what's the reasoning for needing more than the head ref? `group: ${{ github.head_ref }}`

Groups cross workflows, so the per-workflow name is required to avoid concurrency blocking across workflows ; the instances where we use the SHA hash is to ensure multiple runs on the same workflow can happen; the instances where the matrix target is also included is when we want both sides of the target to run.

(never used this before, but that was my understanding when i reviewed)

**Comment by ouchadam** [comment]
thanks for the explanation!

**Review by ouchadam — APPROVED** [review]

**Pull request merged**